File: src/com/mishiranu/instantimage/async/LoadImageTask.java
package com.mishiranu.instantimage.async;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;

import javax.net.ssl.SSLException;

import android.content.Context;

import com.mishiranu.instantimage.R;
import com.mishiranu.instantimage.util.FileManager;
import com.mishiranu.instantimage.util.Preferences;

public class LoadImageTask extends AsyncManager.SimpleTask<Void, Void, Void>
{
	private final Context mContext;
	private final String mUriString;
	
	private int mErrorMessageId;
	private int mContentId;
	private String mMimeType;
	
	public LoadImageTask(Context context, String uriString)
	{
		mContext = context.getApplicationContext();
		mUriString = uriString;
	}
	
	@Override
	protected Void doInBackground(Void... params)
	{
		FileManager.cleanup(mContext);
		HttpURLConnection connection = null;
		try
		{
			connection = (HttpURLConnection) new URL(mUriString).openConnection();
			connection.setConnectTimeout(15000);
			connection.setReadTimeout(15000);
			if (connection.getResponseCode() != HttpURLConnection.HTTP_OK)
			{
				mErrorMessageId = R.string.error_invalid_response;
				return null;
			}
			if (isCancelled()) return null;
			InputStream input = connection.getInputStream();
			int contentId = Preferences.nextContentId(mContext);
			FileManager.FileItem fileItem = FileManager.obtainFile(mContext, mUriString, contentId);
			FileOutputStream output = new FileOutputStream(fileItem.file);
			int count;
			byte[] buffer = new byte[16 * 1024];
			while ((count = input.read(buffer)) >= 0 && !isCancelled()) output.write(buffer, 0, count);
			output.close();
			mContentId = contentId;
			mMimeType = fileItem.getMimeType();
		}
		catch (SSLException e)
		{
			mErrorMessageId = R.string.error_ssl;
		}
		catch (IOException e)
		{
			mErrorMessageId = R.string.error_connection;
		}
		finally
		{
			if (connection != null) connection.disconnect();
		}
		return null;
	}
	
	@Override
	protected void onStoreResult(AsyncManager.Holder holder, Void result)
	{
		holder.storeResult(mErrorMessageId, mContentId, mMimeType);
	}
}

File: src/com/mishiranu/instantimage/async/LoadImageTask.scala
package com.mishiranu.instantimage.async

import android.content.Context

import com.mishiranu.instantimage.R
import com.mishiranu.instantimage.util.{FileManager, Preferences}
import com.mishiranu.instantimage.util.ScalaHelpers._

import com.squareup.okhttp.Response

import rx.android.schedulers.AndroidSchedulers
import rx.lang.scala.Observable
import rx.schedulers.Schedulers

import java.io.{FileOutputStream, IOException}

import javax.net.ssl.SSLException

class LoadImageTask(context: Context, uriString: String) extends RxHttp {
  import LoadImageTask._

  private class InvalidResponseException extends Exception

  private def request: Observable[Response] = {
    request(uriString).header("Referer", uriString.replaceAll("(https?://.*?/).*", "$1")).execute()
  }

  private def getImage(response: Response): Array[Byte] = {
    if (response.code != 200) {
      throw new InvalidResponseException
    }
    response.body.bytes
  }

  private def writeToFile(response: Array[Byte]): Result = {
    val contentId = Preferences.nextContentId()
    val fileItem = FileManager.obtainFile(context, uriString, contentId)
    var outputStream: FileOutputStream = null
    try {
      outputStream = new FileOutputStream(fileItem.file)
      outputStream.write(response)
    } finally {
      if (outputStream != null) {
        outputStream.close()
      }
    }
    Result(contentId, fileItem.mimeType, 0)
  }

  private def handleError(t: Throwable): Result = {
    Result(0, null, t match {
      case _: InvalidResponseException => R.string.error_invalid_response
      case _: SSLException => R.string.error_ssl
      case _: IOException => R.string.error_connection
      case _ => R.string.error_unknown
    })
  }

  def load: Observable[Result] = {
    request.subscribeOn(Schedulers.io)
      .map(getImage)
      .map(writeToFile)
      .onErrorReturn(handleError)
      .observeOn(AndroidSchedulers.mainThread)
  }
}

object LoadImageTask {
  case class Result(contentId: Int, mimeType: String, errorMessageId: Int)

  def apply(context: Context, uriStrings: Iterable[String]): Observable[List[Result]] = {
    Observable.combineLatest(uriStrings.map(s => new LoadImageTask(context.getApplicationContext, s).load))(_.toList)
  }
}
